registry/cache.go

```diff
@@ -252,16 +334,22 @@ func (rc *registryCache) PushBlob(ctx context.Context, repo string, desc ociregi
 }
 
 func (rc *registryCache) MountBlob(ctx context.Context, fromRepo, toRepo string, digest ociregistry.Digest) (ociregistry.Descriptor, error) {
-	// TODO this does *not* need to lock the entire cache during the upstream push (but it *would* be good to block pushing to this specific digest)
-	rc.mu.Lock()
-	defer rc.mu.Unlock()
+	// TODO technically we should also be able to safely imply that "fromRepo" has digest here too (assuming MountBlob success), but need to double check whether the contract of the MountBlob API in OCI is such that it's legal for it to return success if "toRepo" already has "digest" (even if "fromRepo" doesn't)
```

Comment:
At least for Docker Hub, we sure could assume this!

```console
$ crane blob tianon/test@sha256:25be82253336f0b8c4347bc4ecbbcdc85d0e0f118ccf8dc2e119c0a47a0a486e > /dev/null
$ crane blob tianon/scratch@sha256:25be82253336f0b8c4347bc4ecbbcdc85d0e0f118ccf8dc2e119c0a47a0a486e > /dev/null
Error: fetching blob tianon/scratch@sha256:25be82253336f0b8c4347bc4ecbbcdc85d0e0f118ccf8dc2e119c0a47a0a486e: GET https://index.docker.io/v2/tianon/scratch/blobs/sha256:25be82253336f0b8c4347bc4ecbbcdc85d0e0f118ccf8dc2e119c0a47a0a486e: BLOB_UNKNOWN: blob unknown to registry; sha256:25be82253336f0b8c4347bc4ecbbcdc85d0e0f118ccf8dc2e119c0a47a0a486e
$ curl -H "$(crane auth token --header --push --mount tianon/scratch tianon/test)" 'https://registry-1.docker.io/v2/tianon/test/blobs/uploads/?mount=sha256:25be82253336f0b8c4347bc4ecbbcdc85d0e0f118ccf8dc2e119c0a47a0a486e&from=tianon/scratch'
{"errors":[{"code":"BLOB_UPLOAD_UNKNOWN","message":"blob upload unknown to registry","detail":"blob upload unknown"}]}
```

(but whether it's _generally_ safe to assume is not guaranteed by the OCI specification, and in fact the opposite might actually be true -- `from` is allowed to be optional :smile:  https://github.com/opencontainers/distribution-spec/blob/v1.1.0/spec.md#mounting-a-blob-from-another-repository)